src/core/MOM_stoch_eos.F90

```diff
@@ -0,0 +1,212 @@
+!> Provides the ocean stochastic equation of state
+module MOM_stoch_eos
+! This file is part of MOM6. See LICENSE.md for the license.
+use MOM_grid,            only : ocean_grid_type
+use MOM_hor_index,       only : hor_index_type
+use MOM_file_parser,     only : get_param, param_file_type
+use MOM_random,          only : PRNG,random_2d_constructor,random_2d_norm
+use MOM_time_manager,    only : time_type
+use MOM_io,              only : vardesc, var_desc
+use MOM_restart,         only : MOM_restart_CS,is_new_run
+use MOM_diag_mediator,   only : register_diag_field,post_data,diag_ctrl,safe_alloc_ptr
+use MOM_variables,       only : thermo_var_ptrs
+use MOM_verticalGrid,    only : verticalGrid_type
+use MOM_restart,         only : register_restart_field
+use MOM_isopycnal_slopes,only : vert_fill_TS
+!use random_numbers_mod, only : getRandomNumbers,initializeRandomNumberStream,randomNumberStream
+
+implicit none
+#include <MOM_memory.h>
+
+public MOM_stoch_eos_init
+public MOM_stoch_eos_run
+public MOM_calc_varT
+
+real,private ALLOCABLE_, dimension(NIMEM_,NJMEM_) :: l2_inv
+                        !< One over sum of the T cell side side lengths squared
+real,private ALLOCABLE_, dimension(NIMEM_,NJMEM_) :: rgauss !< nondimensional random Gaussian
+real, parameter,private :: tfac=0.27 !< Nondimensional decorrelation time factor, ~1/3.7
+real, parameter,private :: amplitude=0.624499 !< Nondimensional std dev of Gaussian
+integer        ,private :: seed !< PRNG seed
+type(PRNG)  ::  rn_CS !< PRNG control structure
+
+!> Describes parameters of the stochastic component of the EOS
+!! correction, described in Stanley et al. JAMES 2020.
+type, public :: MOM_stoch_eos_CS
```

Comment:
Please consider making this type opaque, by changing this to `type, public :: MOM_stoch_eos_CS ; private`.  The Fortran standard (and the compilers we use) allow for public elements in otherwise private types.  Using opaque types makes it much easier to figure out where in the code things are being set or used.

Reply:
Without further modifications, adding ``private`` to the `MOM_stoch_eos_CS` control structure will break the model. This is because, in the current implementation, subroutine `step_MOM_dynamics` needs access to [use_stoch_eos](https://github.com/gustavo-marques/MOM6/blob/dev-ncar-main-candidate-2022-06-03/src/core/MOM.F90#L1069), [stanley_coeff](https://github.com/gustavo-marques/MOM6/blob/dev-ncar-main-candidate-2022-06-03/src/core/MOM.F90#L1072), and [some](https://github.com/gustavo-marques/MOM6/blob/dev-ncar-main-candidate-2022-06-03/src/core/MOM.F90#L1243-L1245) of the diagnostic ids.

Is addressing this a requirement for merging this PR? 
cc'ing @iangrooms

Reply:
In the interests of moving ahead with this PR, I think that we can accept the PR as whole and deal with this problem later. If we decide to go this route, I will be putting in a formal description of this issue of non-standard code constructs in MOM_stoch_eos.F90 so that we don't lose track of it.